packages/ai/src/workflows/Workflow.ts

```diff
@@ -211,6 +216,38 @@ export class Workflow<Shape extends IOShape = IOShape> {
     });
   }
 
+  addRule<TemplateShape extends IOShape, ProduceShape extends IOShape>({
+    after: template,
+    guard,
+    produce,
+  }: {
+    after: LLMStepTemplate<TemplateShape>;
+    guard: (outputs: Partial<Outputs<TemplateShape>>) => boolean;
+    produce: (
+      outputs: Partial<Outputs<TemplateShape>>
+    ) => [LLMStepTemplate<ProduceShape>, Inputs<ProduceShape>];
```

Comment:
There's a tradeoff here between "make produce more flexible" (i.e. allow it to decide whether to return or not, or return multiple steps) and type safety; I'll probably do a few more experiments on how to make this API nice.